# Lab 2

In [18]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv('s3://ia340lab1-bucket-porter/Real_Estate_Sales_2001-2023_GL.csv') # load the data from s3
df.head() #show the top 5 rows


Serial Number  List Year Date Recorded     Town          Address  \
0        2020177       2020    04/14/2021  Ansonia    323 BEAVER ST   
1        2020225       2020    05/26/2021  Ansonia   152 JACKSON ST   
2        2020348       2020    09/13/2021  Ansonia  230 WAKELEE AVE   
3        2020090       2020    12/14/2020  Ansonia      57 PLATT ST   
4         210288       2021    06/20/2022     Avon   12 BYRON DRIVE   

   Assessed Value  Sale Amount Sales Ratio Property Type Residential Type  \
0        133000.0     248400.0      0.5354   Residential    Single Family   
1        110500.0     239900.0      0.4606   Residential     Three Family   
2        150500.0     325000.0       0.463    Commercial              NaN   
3        127400.0     202500.0      0.6291   Residential       Two Family   
4        179990.0     362500.0      0.4965   Residential            Condo   

  Non Use Code Assessor Remarks OPM remarks  \
0          NaN              NaN         NaN   
1          NaN              NaN         NaN   
2          NaN              NaN         NaN   
3          NaN              NaN         NaN   
4          NaN              NaN         NaN   

                             Location  
0          POINT (-73.06822 41.35014)  
1                                 NaN  
2                                 NaN  
3                                 NaN  
4  POINT (-72.879115982 41.773452988)

In [19]:
df.shape #shows the number of rows and columns

(1141722, 14)

In [20]:
df.info() #shows basic info of data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1141722 entries, 0 to 1141721
Data columns (total 14 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   Serial Number     1141722 non-null  int64  
 1   List Year         1141722 non-null  int64  
 2   Date Recorded     1141720 non-null  object 
 3   Town              1141722 non-null  object 
 4   Address           1141671 non-null  object 
 5   Assessed Value    1141722 non-null  float64
 6   Sale Amount       1141722 non-null  float64
 7   Sales Ratio       1141722 non-null  object 
 8   Property Type     759276 non-null   object 
 9   Residential Type  738804 non-null   object 
 10  Non Use Code      324807 non-null   object 
 11  Assessor Remarks  181090 non-null   object 
 12  OPM remarks       14346 non-null    object 
 13  Location          341241 non-null   object 
dtypes: float64(2), int64(2), object(10)
memory usage: 121.9+ MB


In [21]:
df.describe()

Serial Number     List Year  Assessed Value   Sale Amount
count   1.141722e+06  1.141722e+06    1.141722e+06  1.141722e+06
mean    5.457366e+05  2.011673e+03    2.833275e+05  4.104510e+05
std     7.450110e+06  7.018679e+00    1.656128e+06  5.048996e+06
min     0.000000e+00  2.001000e+03    0.000000e+00  0.000000e+00
25%     3.090300e+04  2.005000e+03    8.991000e+04  1.461000e+05
50%     9.016000e+04  2.012000e+03    1.419800e+05  2.375000e+05
75%     1.801550e+05  2.018000e+03    2.300600e+05  3.837500e+05
max     2.000500e+09  2.023000e+03    8.815100e+08  5.000000e+09

There are 1141722 rows and 14 columns. 2 columns are floats, 2 integers and 10 objects.

In [49]:
df = df.rename(columns={'List Year': 'Year Listed'})

In [24]:
df.isnull().sum()

Serial Number             0
List Year                 0
Date Recorded             2
Town                      0
Address                  51
Assessed Value            0
Sale Amount               0
Sales Ratio               0
Property Type        382446
Residential Type     402918
Non Use Code         816915
Assessor Remarks     960632
OPM remarks         1127376
Location             800481
dtype: int64

In [37]:
df = df.dropna()

 I changed List Year to Year Listed and using dropna and isnull I found and removed all but 720 rows due to missing values.

In [39]:
df['Town']

674             Berlin
715          Fairfield
1381           Danbury
1438           Danbury
2066       East Granby
              ...     
1140329        Shelton
1140805        Shelton
1140826       Portland
1141238        Redding
1141721        Shelton
Name: Town, Length: 720, dtype: object

In [56]:
df.sort_values(by='Sale Amount')

Serial Number  Year Listed Date Recorded        Town  \
1112366         230079         2023    11/21/2023      Groton   
1105173         230080         2023    11/21/2023      Groton   
935781          180148         2018    04/29/2019     Clinton   
1087034         220809         2022    08/07/2023     Norwich   
1097268         220807         2022    08/07/2023     Norwich   
...                ...          ...           ...         ...   
1055467         220511         2022    04/26/2023    Branford   
1085483         220311         2022    12/20/2022     Norwalk   
8940            200403         2020    07/02/2021   East Lyme   
1090670         220371         2022    01/13/2023     Norwalk   
64643         20200078         2020    07/06/2021  Willington   

                          Address  Assessed Value  Sale Amount Sales Ratio  \
1112366     268 FLANDERS RD LOT 1          9450.0       2000.0       4.725   
1105173    268 FLANDERS RD LOT 21          6230.0       2500.0       2.492   
935781   133 WEST MAIN ST LOT C-4         33400.0       4500.0    7.422222   
1087034        44 NORWICH AVE #44         10600.0       5000.0        2.12   
1097268        44 NORWICH AVE #21          8700.0       5000.0        1.74   
...                           ...             ...          ...         ...   
1055467              GREEN ISLAND        769900.0    3500000.0        0.22   
1085483          51 PINE POINT RD       1000520.0    3675000.0      0.2723   
8940          249 OLD BLACK PT RD       2083410.0    6568750.0     0.31717   
1090670        11 HARBOR BLUFF LN       1711290.0    7000000.0      0.2445   
64643              224 RIVER ROAD        223070.0  318790019.0      0.0007   

         Property Type Residential Type             Non Use Code  \
1112366    Residential    Single Family               25 - Other   
1105173    Residential    Single Family               25 - Other   
935781   Single Family    Single Family               25 - Other   
1087034    Residential    Single Family               25 - Other   
1097268    Residential    Single Family               25 - Other   
...                ...              ...                      ...   
1055467    Residential    Single Family               25 - Other   
1085483    Residential    Single Family  07 - Change in Property   
8940       Residential    Single Family  07 - Change in Property   
1090670    Residential            Condo  07 - Change in Property   
64643      Residential    Single Family               25 - Other   

                                          Assessor Remarks  \
1112366          MOBILE HOME - SENDING SALES QUESTIONNAIRE   
1105173                        SENDING SALES QUESTIONNAIRE   
935781                      MOBILE HOME ONLY ON LEASED LOT   
1087034                                              OTHER   
1097268                                              OTHER   
...                                                    ...   
1055467  PURCHASED GREEN ISLAND IN ONE DEED, OTHER PARC...   
1085483                                          6-57A-1-0   
8940                                            WATERFRONT   
1090670                              6-16C-2-11; CONDO PUD   
64643                                             COLONIAL   

                                               OPM remarks  \
1112366    RATIO - PURCHASE FROM OWNER OF MOBILE HOME PARK   
1105173    RATIO - PURCHASE FROM OWNER OF MOBILE HOME PARK   
935781                                BELOW MARKET - RATIO   
1087034                         MOBILE HOME - BELOW MARKET   
1097268                         MOBILE HOME - BELOW MARKET   
...                                                    ...   
1055467  PER MLS SALE INCLUDES 78 ACRES OF OYSTER BEDS ...   
1085483  ADDITIONS AND ALTERATIONS, NEW ASSESSMENT = $1...   
8940     PER MLS HOUSE COMPLETED JAN. 2020 - ASSESSMENT...   
1090670  PER MLS HOME HAS 5 BEDROOMS, 5.5 BATHS AND 2,0...   
64643                        INCORRECT SALE PRI

In [58]:
df[0:5]

Serial Number  Year Listed Date Recorded         Town  \
674          210296         2021    07/06/2022       Berlin   
715          210197         2021    11/30/2021    Fairfield   
1381         211177         2021    06/28/2022      Danbury   
1438         210856         2021    04/13/2022      Danbury   
2066         210140         2021    09/15/2022  East Granby   

                        Address  Assessed Value  Sale Amount Sales Ratio  \
674            355 RESERVOIR RD        172300.0     240000.0    0.717917   
715           125 VESPER STREET        264320.0     334000.0    0.791377   
1381      55 MILL PLAIN RD 13-1        146600.0     285000.0    0.514386   
1438             60 NABBY RD 78        151600.0     350000.0    0.433143   
2066  143-145 NORTH MAIN STREET        130600.0      75000.0    1.741333   

     Property Type Residential Type             Non Use Code  \
674    Residential    Single Family               25 - Other   
715    Residential    Single Family               25 - Other   
1381   Residential            Condo               25 - Other   
1438   Residential            Condo               25 - Other   
2066   Residential       Two Family  07 - Change in Property   

                Assessor Remarks  \
674                        OTHER   
715              POSSIBLE DURESS   
1381                   D15004-80   
1438                   L08054-78   
2066  PP TOO LOW, RATIO TOO HIGH   

                                            OPM remarks  \
674                  TWO SALES IN ONE DAY - SEE #210295   
715   PER MLS SLOPING ISSUES ON 2ND FLOOR. ENGINEERS...   
1381                 DUPLICATE ENTRY - ALSO SEE #211176   
1438                 DUPLICATE ENTRY - ALSO SEE #210855   
2066  POOR CONDITION PER MLS - ASSESSMENT REDUCED TO...   

                                Location  
674   POINT (-72.804847962 41.612923994)  
715   POINT (-73.234403987 41.179839983)  
1381  POINT (-73.502786995 41.391933021)  
1438  POINT (-73.421529975 41.429339988)  
2066  POINT (-72.720844025 41.963067982)

In [57]:
df[(df['Year Listed']>= 2020)]

Serial Number  Year Listed Date Recorded         Town  \
674             210296         2021    07/06/2022       Berlin   
715             210197         2021    11/30/2021    Fairfield   
1381            211177         2021    06/28/2022      Danbury   
1438            210856         2021    04/13/2022      Danbury   
2066            210140         2021    09/15/2022  East Granby   
...                ...          ...           ...          ...   
1140329         230355         2023    05/28/2024      Shelton   
1140805         230398         2023    06/21/2024      Shelton   
1140826         230007         2023    10/16/2023     Portland   
1141238        2300040         2023    01/17/2024      Redding   
1141721         230383         2023    06/06/2024      Shelton   

                           Address  Assessed Value  Sale Amount Sales Ratio  \
674               355 RESERVOIR RD        172300.0     240000.0    0.717917   
715              125 VESPER STREET        264320.0     334000.0    0.791377   
1381         55 MILL PLAIN RD 13-1        146600.0     285000.0    0.514386   
1438                60 NABBY RD 78        151600.0     350000.0    0.433143   
2066     143-145 NORTH MAIN STREET        130600.0      75000.0    1.741333   
...                            ...             ...          ...         ...   
1140329          208 SOUNDVIEW AVE        267750.0     275500.0      0.9719   
1140805              36 SALLYS WAY        149310.0     529900.0      0.2818   
1140826             11 WILDWOOD RD        155400.0     225000.0      0.6907   
1141238           358 NEWTOWN TPKE        700000.0    1325000.0      0.5283   
1141721           512 ANTELOPE TRL        306460.0     659813.0      0.4645   

        Property Type Residential Type             Non Use Code  \
674       Residential    Single Family               25 - Other   
715       Residential    Single Family               25 - Other   
1381      Residential            Condo               25 - Other   
1438      Residential            Condo               25 - Other   
2066      Residential       Two Family  07 - Change in Property   
...               ...              ...                      ...   
1140329   Residential    Single Family               25 - Other   
1140805   Residential            Condo  07 - Change in Property   
1140826   Residential    Single Family               25 - Other   
1141238   Residential    Single Family               25 - Other   
1141721   Residential            Condo               25 - Other   

                   Assessor Remarks  \
674                           OTHER   
715                 POSSIBLE DURESS   
1381                      D15004-80   
1438                      L08054-78   
2066     PP TOO LOW, RATIO TOO HIGH   
...                             ...   
1140329                      125 58   
1140805                    51 13 36   
1140826                    042-0027   
1141238                  RELOCATION   
1141721                   89 33 512   

                                               OPM remarks  \
674                     TWO SALES IN ONE DAY - SEE #210295   
715      PER MLS SLOPING ISSUES ON 2ND FLOOR. ENGINEERS...   
1381                    DUPLICATE ENTRY - ALSO SEE #211176   
1438                    DUPLICATE ENTRY - ALSO SEE #210855   
2066     POOR CONDITION PER MLS - ASSESSMENT REDUCED TO...   
...                                                    ...   
1140329                                 SHORT SALE PER MLS   
1140805                    NEW CONSTRUCTION - CO 6/12/2024   
1140826  BELOW MARKET - RESOLD 12/19/2023 $300,500 WITH...   
1141238                                       SEE #2300039   
1141721                   PER MLS CLOSING PRICE = $675,000   

                                   Location  
674      POINT (-72.804847962 41.612923994)  
715      POINT (-73.234403987 41.179839983)  
1381     POINT (-73.502786995 41.391933021)  
1438     POINT (-73.421529975 41.429339988)  
2066     POINT (-7